test/e2e/metriconly/metrics_test.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_metric_only
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/node-problem-detector/pkg/util/metrics"
+	"k8s.io/node-problem-detector/test/e2e/lib/gce"
+	"k8s.io/node-problem-detector/test/e2e/lib/npd"
+
+	"github.com/avast/retry-go"
+	"github.com/onsi/ginkgo"
+	"github.com/pborman/uuid"
+)
+
+var _ = ginkgo.Describe("NPD should export Prometheus metrics for detected problems", func() {
+	var instance gce.Instance
+
+	ginkgo.BeforeEach(func() {
+		var err error
+		// TODO(xueweiz): Creating instance for each test case is slow. We should either reuse the instance
+		// between tests, or have a way to run these tests in parallel.
+		instance, err = gce.CreateInstance(
+			gce.Instance{
+				Name:           "npd-metrics-" + *image + "-" + uuid.NewUUID().String()[:8],
+				Zone:           *zone,
+				Project:        *project,
+				SshKey:         *sshKey,
+				SshUser:        *sshUser,
+				ComputeService: computeService,
+			},
+			*image,
+			*imageProject)
+		if err != nil {
+			ginkgo.Fail(fmt.Sprintf("Unable to create test instance: %v", err))
+		}
+
+		err = npd.SetupNPD(instance, *npdBuildTar)
+		if err != nil {
+			ginkgo.Fail(fmt.Sprintf("Unable to setup NPD: %v", err))
+		}
+	})
+
+	ginkgo.Context("when started on a clean instance,", func() {
+
+		ginkgo.It("NPD should export Prometheus metrics for uptime", func() {
+			var gotMetrics []metrics.Float64MetricRepresentation
+			var err error
+
+			verifyUptimeMetric := func() error {
+				gotMetrics, err = npd.FetchNPDMetrics(instance)
+				if err != nil {
+					return fmt.Errorf("Error fetching NPD metrics: %v", err)
+				}
+
+				_, err = metrics.GetFloat64Metric(gotMetrics, "host_uptime", map[string]string{}, false)
+				if err != nil {
+					return fmt.Errorf("Failed to find uptime metric: %v.\nHere is all NPD exported metrics: %v",
+						err, gotMetrics)
+				}
+				return nil
+			}
+
+			// Wait for NPD to be ready for a maximum of 120 seconds.
+			err = retry.Do(verifyUptimeMetric,
+				retry.Delay(10*time.Second),
+				retry.Attempts(12),
+				retry.DelayType(retry.FixedDelay))
+
+			if err != nil {
+				ginkgo.Fail(fmt.Sprintf("Failed to find uptime metric: %v.\nHere is all NPD exported metrics: %v",
+					err, gotMetrics))
+			}
+		})
+	})
+
+	ginkgo.AfterEach(func() {
+		npdMetrics := instance.RunCommand("curl http://localhost:20257/metrics")
```

Comment:
Is this and below printing journal log for debugging? Would it make the output look too verbose?

Reply:
Yes, these prints are for debugging. I think it's reasonable to print out the NPD reported metrics. Because after all, they are the key subject that we are asserting against (e.g. NPD has this metric and that metric...).

In my opinion, the output looks OK for now. See example at https://gubernator-internal.googleplex.com/build/gke-prow/logs/ci-npd-e2e-cos-experimental/1161702249425014784/

I printed out the journal logs, mainly to catch invalid config file problem. i.e. The case where someone changed some config / arguments and made NPD panic just when starting.

However, I totally agree that it's best if we do not print out journal logs. They are not the tested subject, and are merely used for debugging the tested subject. And most importantly, they CAN potentially be large when the test gets more complicated.

I think the best practice for uploading large debugging text in testing, is to tar these texts and upload them to some test artifact storage. Do you know whether this can be done with Prow? Can you point me to some code sample?

Reply:
How about putting those into artifacts? For example, current NPD k8s e2e test stores node-problem-detector.log as one of the artifact. See the following artifacts from a CI job.
https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-npd-e2e-kubernetes-gce-gci/1161681249476022272/artifacts/e2e-3e359a827e-7abb6-minion-group-7pkc/

@krzyzacy may know how to set that up.

Reply:
Thanks!
I believe this is how the artifact was setup:
https://github.com/kubernetes/test-infra/blob/master/prow/pod-utilities.md#what-the-test-container-can-expect

Experimenting now.

Reply:
Nice it worked :)
See this example: https://gubernator-internal.googleplex.com/build/gke-prow/logs/ci-npd-e2e-cos-experimental/1161754845946843136/

And a screenshot: 
![image](https://user-images.githubusercontent.com/15825828/63058716-44d06c00-bea2-11e9-869c-326bd3b76ec8.png)